core/shared/src/main/scala/zio/syntax/IOSyntax.scala

```diff
@@ -39,21 +39,23 @@ object IOSyntax {
     def collectAll: IO[E, List[A]]                     = IO.collectAll(ios)
   }
 
-  final class IOTuple2[E, A, B](val ios2: (IO[E, A], IO[E, B])) extends AnyVal {
-    def map2[C](f: (A, B) => C): IO[E, C] = ios2._1.flatMap(a => ios2._2.map(f(a, _)))
+  final class IOTuple2[E, RA, A, RB, B](val ios2: (ZIO[RA, E, A], ZIO[RB, E, B])) extends AnyVal {
```

Comment:
Would this widen correctly to the LUB error type if left and right IO in a tuple have different error types?

Reply:
I'm not sure... if it already did that, it should still do that. This could probably further modified to correctly infer the error LUB, if it doesn't already, but this PR is just for allowing environment types.

Reply:
Yes, it should infer correctly.

Reply:
I tried it out and it does indeed infer the error LUB:

```scala
scala> import zio.syntax._
import zio.syntax._

scala> val a = ZIO(10)
a: zio.Task[Int] = zio.ZIO$EffectPartial@29dd4b0b

scala> val b = ZIO.succeed("hello")
b: zio.UIO[String] = zio.ZIO$Succeed@45db18dd

scala> val c = ZIO(20).refineOrDie { case err: Throwable => new IllegalArgumentException("Oh no!", err) }
c: zio.ZIO[Any,IllegalArgumentException,Int] = <function1>

scala> val all = (a, b, c).map3 {
     |   (i1, str, i2) => i1 + str.length + i2
     | }
all: zio.ZIO[Any with Any with Any,Throwable,Int] = zio.ZIO$FlatMap@363a9244
```

If the `Any with Any with Any` is bothersome, it can be solved with an additional implicit.